File: TrainTicketsSystem.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.HashMap;
import java.util.Map;

public class Main {
    public static void main(String[] args) throws IOException {
        BufferedReader reader = new BufferedReader(new InputStreamReader(System.in));
        int n = Integer.parseInt(reader.readLine());
        HashMap<String, Double> prices = new HashMap<>();
        HashMap<String, Double> profit = new HashMap<>();
        for (int i = 0; i < n; i++){
            String input = reader.readLine();
            String[] data = input.substring(0, input.length() - 3).split(" ");
            prices.put(data[0], Double.parseDouble(data[1]));
        }
        for (Map.Entry<String, Double> entry : prices.entrySet()){
            String k = entry.getKey();
            double v = entry.getValue();
            System.out.println(k + ":");
            String input = reader.readLine();
            int i = 0;
            double profitPerCity = 0.0;
            while (!input.equals("NEXT DESTINATION")){
                String ageStr = "";
                for (char c : input.toCharArray()){
                    if (c == 'y'){
                        break;
                    }
                    ageStr += c;
                }
                int age = Integer.parseInt(ageStr);
                int time = Integer.parseInt(input.substring(input.length() - 5, input.length()-3));
                double percantage = 0.0;
                if ((time >= 0 && time < 7) || (time >= 19 && time <= 23)){
                    percantage += 0.05;
                }
                if (age >= 7 && age < 12){
                    percantage += 0.2;
                } else if (age >= 12 && age < 18 || age > 64){
                    percantage += 0.1;
                }
                double price = v;
                double discount = price*percantage;
                price -= discount;
                if (age < 7){
                    price = 0;
                    i--;
                }
                profitPerCity += price;
                i++;
                input = reader.readLine();
            }
            if (i >= 4){
                profitPerCity -= prices.get(k) * 0.05 * i;
            }
            profit.put(k, profitPerCity);
        }
        profit.forEach((k, v) -> {
            System.out.println(k + " " + v);
        });
    }
}

File: train_tickets_system.rs
use std::collections::HashMap;
use std::io;
fn calculate_discount(age: u32, time: u32, price: &f64) -> Result<f64, &str> {
    let mut discount = 0.0;
    if time < 7 || (time >= 19 && time <= 23) {
        discount += 0.05;
    }
    if age >= 7 && age < 12 {
        discount += 0.2;
    } else if age >= 12 && age < 18 || age > 64 {
        discount += 0.1;
    }
    if age > 6 { Ok(discount * price) } else { Err("") }
}
fn calculate_profit(price: &f64) -> f64 {
    let mut i = 0;
    let mut profit = 0.0;
    loop {
        let mut input = String::new();
        io::stdin().read_line(&mut input).unwrap();
        let input = input.trim().to_string();
        if input == "NEXT DESTINATION" {
            break;
        }
        let age: u32 = input[0..input.find('y').unwrap()].parse().unwrap();
        let time: u32 = input[input.len() - 5..input.len() - 3].parse().unwrap();
        let discount = calculate_discount(age, time, &price);
        match discount {
            Ok(discount) => {
                profit += price - discount;
                i += 1;
            }
            Err(_) => (),
        }
    }
    if i >= 4 {
        profit -= price * 0.05 * i as f64;
    }
    profit
}

fn main() {
    let mut input = String::new();
    io::stdin().read_line(&mut input).unwrap();
    let n = input.trim().parse().unwrap();
    let mut prices = HashMap::new();
    let mut profits = HashMap::new();
    for _ in 0..n {
        let mut input = String::new();
        io::stdin().read_line(&mut input).unwrap();
        let input = input.trim()[..input.len() - 4].to_string();
        let data: Vec<String> = input.split(" ").map(|s| s.to_string()).collect();
        let price: f64 = data[1].parse().unwrap();
        prices.insert(data[0].to_string(), price);
    }
    for (k, v) in prices {
        println!("{}:", k);
        let total_price = calculate_profit(&v);
        profits.insert(k, total_price);
    }
    for (k, v) in profits {
        println!("{} : {}", k, v);
    }
}
